### PR TITLE
Typing overload

### PR DESCRIPTION
**Pull request by janrito**
Resolves https://github.com/UKPLab/sentence-transformers/issues/2758

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hello!

## SentenceTransformer
I think this is a very good start, but I'm missing some edge cases, e.g. convert_to_numpy and convert_to_tensor both True:
image

These are all the cases that we've got:
```
sentences: str
    convert_to_numpy: False
        convert_to_tensor: False -> output: torch.Tensor (1)
        convert_to_tensor: True -> output: torch.Tensor (2)
    convert_to_numpy: True
        convert_to_tensor: False -> output: numpy.ndarray (3)
        convert_to_tensor: True -> output: torch.Tensor (4)

sentences: List[str]
    convert_to_numpy: False
        convert_to_tensor: False -> output: List[torch.Tensor] (5)
        convert_to_tensor: True -> output: torch.Tensor (6)
    convert_to_numpy: True
        convert_to_tensor: False -> output: numpy.ndarray (7)
        convert_to_tensor: True -> output: torch.Tensor (8)
```

The minimum number of overloads to cover these cases is 4:
* 1 (Tensor)
* 2, 4, 6, 8 (Tensor)
* 3, 7 (ndarray)
* 5 (List[torch.Tensor])

I verified that my changes worked locally through this:

```python
from typing import List
from sentence_transformers import SentenceTransformer
import numpy as np
import torch

model = SentenceTransformer("all-MiniLM-L6-v2")

output1: torch.Tensor = model.encode("This is a string", convert_to_numpy=False, convert_to_tensor=False)
output2: torch.Tensor = model.encode("This is a string", convert_to_numpy=False, convert_to_tensor=True)
output3: np.ndarray = model.encode("This is a string", convert_to_numpy=True, convert_to_tensor=False)
output4: torch.Tensor = model.encode("This is a string", convert_to_numpy=True, convert_to_tensor=True)
output5: List[torch.Tensor] = model.encode(["This is a string", "This is another string"], convert_to_numpy=False, convert_to_tensor=False)
output6: torch.Tensor = model.encode(["This is a string", "This is another string"], convert_to_numpy=False, convert_to_tensor=True)
output7: np.ndarray = model.encode(["This is a string", "This is another string"], convert_to_numpy=True, convert_to_tensor=False)
output8: torch.Tensor = model.encode(["This is a string", "This is another string"], convert_to_numpy=True, convert_to_tensor=True)
```

and 
```
python -m mypy .\demo_inference.py
```

---

## CrossEncoder

We actually have roughly the same setup for CrossEncoders:
```
sentences: List[str]
    convert_to_numpy: False
        convert_to_tensor: False -> output: torch.Tensor (1)
        convert_to_tensor: True -> output: torch.Tensor (2)
    convert_to_numpy: True
        convert_to_tensor: False -> output: numpy.ndarray (3)
        convert_to_tensor: True -> output: torch.Tensor (4)

sentences: List[List[str]]
    convert_to_numpy: False
        convert_to_tensor: False -> output: List[torch.Tensor] (5)
        convert_to_tensor: True -> output: torch.Tensor (6)
    convert_to_numpy: True
        convert_to_tensor: False -> output: numpy.ndarray (7)
        convert_to_tensor: True -> output: torch.Tensor (8)
```

I see that the old typing didn't even cover the `List[str]` case, but we can add that now. We again need the 4 overloads. Now we can verify that it works with:

```python
from typing import List
from sentence_transformers import CrossEncoder
import numpy as np
import torch

model = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")

output1a: torch.Tensor = model.predict(("This is a test", "This is another test"), convert_to_numpy=False, convert_to_tensor=False)
output1b: torch.Tensor = model.predict(["This is a test", "This is another test"], convert_to_numpy=False, convert_to_tensor=False)
output2a: torch.Tensor = model.predict(("This is a test", "This is another test"), convert_to_numpy=False, convert_to_tensor=True)
output2b: torch.Tensor = model.predict(["This is a test", "This is another test"], convert_to_numpy=False, convert_to_tensor=True)
output3a: np.ndarray = model.predict(("This is a test", "This is another test"), convert_to_numpy=True, convert_to_tensor=False)
output3b: np.ndarray = model.predict(["This is a test", "This is another test"], convert_to_numpy=True, convert_to_tensor=False)
output4a: torch.Tensor = model.predict(("This is a test", "This is another test"), convert_to_numpy=True, convert_to_tensor=True)
output4b: torch.Tensor = model.predict(["This is a test", "This is another test"], convert_to_numpy=True, convert_to_tensor=True)
output5a: List[torch.Tensor] = model.predict([("This is a test", "This is another test"), ("This is another test", "This is a test")], convert_to_numpy=False, convert_to_tensor=False)
output5b: List[torch.Tensor] = model.predict([["This is a test", "This is another test"], ["This is another test", "This is a test"]], convert_to_numpy=False, convert_to_tensor=False)
output6a: torch.Tensor = model.predict([("This is a test", "This is another test"), ("This is another test", "This is a test")], convert_to_numpy=False, convert_to_tensor=True)
output6b: torch.Tensor = model.predict([["This is a test", "This is another test"], ["This is another test", "This is a test"]], convert_to_numpy=False, convert_to_tensor=True)
output7a: np.ndarray = model.predict([("This is a test", "This is another test"), ("This is another test", "This is a test")], convert_to_numpy=True, convert_to_tensor=False)
output7b: np.ndarray = model.predict([["This is a test", "This is another test"], ["This is another test", "This is a test"]], convert_to_numpy=True, convert_to_tensor=False)
output8a: torch.Tensor = model.predict([("This is a test", "This is another test"), ("This is another test", "This is a test")], convert_to_numpy=True, convert_to_tensor=True)
output8b: torch.Tensor = model.predict([["This is a test", "This is another test"], ["This is another test", "This is a test"]], convert_to_numpy=True, convert_to_tensor=True)
```


and 
```
python -m mypy .\demo_inference.py
```

Both give:

```
Success: no issues found in 1 source file
```

I think this should be ready for merging now. What do you think?

- Tom Aarsen

**Comment by janrito** [comment]
Yeah, I think this is great. I guess adding the case where `convert_to_numpy` and `convert_to_tensor` reflects the function's behaviour - it will produce a Tensor, and it is valid, if awkward input.

> I see that the old typing didn't even cover the List[str] 
Maybe we can treat this as a a separate issue, but there might be some way of specifying this as a broader type i.e. `Iterable` or `Sequence` - as this method also accepts and works with numpy arrays and pandas serie, etc

**Comment by tomaarsen** [comment]
> as this method also accepts and works with numpy arrays and pandas serie, etc

Fair enough, perhaps `Sequence[str]` is the best option here? Instead of `Union[List[str], Tuple[str, str]]`.

**Review by tomaarsen — CHANGES_REQUESTED** [review]
Not sure about `np.ndarray`, and I was hoping that `Sequence` could make the typing a bit more concise, but that doesn't seem to be the case

**Comment by janrito** [comment]
I was too quick, this doesn't work. Sequence has a few edge cases that I hand't considered. I tried `npt.ArrayLike` which seems to work fine. 

I'll push a commit and then you can decide

**Comment by janrito** [comment]
Here is the test :

```python
from typing import List

import numpy as np
import pandas as pd
import torch
from sentence_transformers import CrossEncoder, SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

output1: torch.Tensor = model.encode(
    "This is a string", convert_to_numpy=False, convert_to_tensor=False
)
output2: torch.Tensor = model.encode(
    "This is a string", convert_to_numpy=False, convert_to_tensor=True
)
output3: np.ndarray = model.encode(
    "This is a string", convert_to_numpy=True, convert_to_tensor=False
)
output4: torch.Tensor = model.encode(
    "This is a string", convert_to_numpy=True, convert_to_tensor=True
)
output5: List[torch.Tensor] = model.encode(
    ["This is a string", "This is another string"],
    convert_to_numpy=False,
    convert_to_tensor=False,
)
output6: torch.Tensor = model.encode(
    ["This is a string", "This is another string"],
    convert_to_numpy=False,
    convert_to_tensor=True,
)
output7: np.ndarray = model.encode(
    ["This is a string", "This is another string"],
    convert_to_numpy=True,
    convert_to_tensor=False,
)
output8: torch.Tensor = model.encode(
    ["This is a string", "This is another string"],
    convert_to_numpy=True,
    convert_to_tensor=True,
)
output9: torch.Tensor = model.encode(
    pd.Series(["This is a string", "This is another string"]),
    convert_to_numpy=True,
    convert_to_tensor=True,
)
output10: torch.Tensor = model.encode(
    np.array(["This is a string", "This is another string"], dtype=str),
    convert_to_numpy=True,
    convert_to_tensor=True,
)


model = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")

output1a: torch.Tensor = model.predict(
    ("This is a test", "This is another test"),
    convert_to_numpy=False,
    convert_to_tensor=False,
)
output1b: torch.Tensor = model.predict(
    ["This is a test", "This is another test"],
    convert_to_numpy=False,
    convert_to_tensor=False,
)
output2a: torch.Tensor = model.predict(
    ("This is a test", "This is another test"),
    convert_to_numpy=False,
    convert_to_tensor=True,
)
output2b: torch.Tensor = model.predict(
    ["This is a test", "This is another test"],
    convert_to_numpy=False,
    convert_to_tensor=True,
)
output3a: np.ndarray = model.predict(
    ("This is a test", "This is another test"),
    convert_to_numpy=True,
    convert_to_tensor=False,
)
output3b: np.ndarray = model.predict(
    ["This is a test", "This is another test"],
    convert_to_numpy=True,
    convert_to_tensor=False,
)
output4a: torch.Tensor = model.predict(
    ("This is a test", "This is another test"),
    convert_to_numpy=True,
    convert_to_tensor=True,
)
output4b: torch.Tensor = model.predict(
    ["This is a test", "This is another test"],
    convert_to_numpy=True,
    convert_to_tensor=True,
)
output5a: List[torch.Tensor] = model.predict(
    [
        ("This is a test", "This is another test"),
        ("This is another test", "This is a test"),
    ],
    convert_to_numpy=False,
    convert_to_tensor=False,
)
output5b: List[torch.Tensor] = model.predict(
    [
        ["This is a test", "This is another test"],
        ["This is another test", "This is a test"],
    ],
    convert_to_numpy=False,
    convert_to_tensor=False,
)
output6a: torch.Tensor = model.predict(
    [
        ("This is a test", "This is another test"),
        ("This is another test", "This is a test"),
    ],
    convert_to_numpy=False,
    convert_to_tensor=True,
)
output6b: torch.Tensor = model.predict(
    [
        ["This is a test", "This is another test"],
        ["This is another test", "This is a test"],
    ],
    convert_to_numpy=False,
    convert_to_tensor=True,
)
output7a: np.ndarray = model.predict(
    [
        ("This is a test", "This is another test"),
        ("This is another test", "This is a test"),
    ],
    convert_to_numpy=True,
    convert_to_tensor=False,
)
output7b: np.ndarray = model.predict(
    [
        ["This is a test", "This is another test"],
        ["This is another test", "This is a test"],
    ],
    convert_to_numpy=True,
    convert_to_tensor=False,
)
output8a: torch.Tensor = model.predict(
    [
        ("This is a test", "This is another test"),
        ("This is another test", "This is a test"),
    ],
    convert_to_numpy=True,
    convert_to_tensor=True,
)
output8b: torch.Tensor = model.predict(
    [
        ["This is a test", "This is another test"],
        ["This is another test", "This is a test"],
    ],
    convert_to_numpy=True,
    convert_to_tensor=True,
)
output9a: torch.Tensor = model.predict(
    pd.DataFrame(
        [
            ["This is a test", "This is another test"],
            ["This is another test", "This is a test"],
        ]
    ),
    convert_to_numpy=True,
    convert_to_tensor=True,
)

output10a: torch.Tensor = model.predict(
    np.array(
        [
            ["This is a test", "This is another test"],
            ["This is another test", "This is a test"],
        ]
    ),
    convert_to_numpy=True,
    convert_to_tensor=True,
)
```

and 
```sh
→ python -m mypy demo_inference.py
[...]
Success: no issues found in 1 source file
```

**Comment by janrito** [comment]
Sequence seems to match a string, which is not what you wouild expect here https://github.com/pandas-dev/pandas-stubs/issues/384

File: sentence_transformers/cross_encoder/CrossEncoder.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 
 import numpy as np
+import numpy.typing as npt
 import torch
 from torch import Tensor, nn
 from torch.optim import Optimizer
 from torch.utils.data import DataLoader
 from tqdm.autonotebook import tqdm, trange
-from transformers import AutoConfig, AutoModelForSequenceClassification, AutoTokenizer, is_torch_npu_available
+from transformers import AutoConfig, AutoModelForListClassification, AutoTokenizer, is_torch_npu_available
```
Comment (tomaarsen) [inline sentence_transformers/cross_encoder/CrossEncoder.py:13]: There seem to be some accidental replaces here

Comment (tomaarsen) [reply sentence_transformers/cross_encoder/CrossEncoder.py]: All good :D

**Comment by janrito** [comment]
What do you think?

**Comment by janrito** [comment]
If you think this requires more thought, I'd be happy to rever these changes and just merge the initial overloads

**Comment by tomaarsen** [comment]
Apologies for the delay, I've indeed been thinking on the best approach. I think perhaps we should remove the `npt.ArrayLike` as a valid input: I don't think we should recommend people to use numpy arrays with string inputs. If we just remove those, then I think this should be ready to go :)

- Tom Aarsen

**Review by tomaarsen — APPROVED** [review]
> Numpy just released a variable length string data type which might make this more palatable in the future [numpy.org/doc/stable//release/2.0.0-notes.html#highlights](https://numpy.org/doc/stable//release/2.0.0-notes.html#highlights)

Good to know!
I'm aware that Numpy reached the 2.0 release, but only because I was forced to restrict Sentence Transformers to numpy<2 for now due to incompatibilities on Windows 😅 

---

Mypy seems to pass still, and I quite like what we've got now I think. I think we're ready to merge when the tests go green, what do you think?

- Tom Aarsen

**Comment by tomaarsen** [comment]
I appreciate your work on this, thank you for taking the time! And for quickly taking care of my reviews/suggestions :)

- Tom Aarsen

**Comment by janrito** [comment]
Thanks @tomaarsen !

**Comment by aelamran** [comment]
Hi guys @janrito @tomaarsen 

I believe encode overloads do not take into account arguments of type Pillow.Image although this is something expected to be used with encode.

Mypy logs:
` No overload variant of "encode" of "SentenceTransformer" matches argument types "Image", "bool" `

Are you planning to cover this?
